pennylane/transforms/qcut.py

```diff
@@ -307,7 +308,28 @@ def fragment_graph(graph: MultiDiGraph) -> Tuple[Tuple[MultiDiGraph], MultiDiGra
 
         communication_graph.add_edge(start_fragment, end_fragment, pair=(node1, node2))
 
-    return subgraphs, communication_graph
+    terminal_indices = [i for i, s in enumerate(subgraphs) for n in measure_nodes if s.has_node(n)]
+
+    subgraphs_connected_to_measurements = []
+    subgraphs_indices_to_remove = []
+    prepare_nodes_removed = []
+
+    for i, s in enumerate(subgraphs):
+        if any(has_path(communication_graph, i, t) for t in terminal_indices):
+            subgraphs_connected_to_measurements.append(s)
+        else:
+            subgraphs_indices_to_remove.append(i)
+            prepare_nodes_removed.extend([n for n in s.nodes if isinstance(n, PrepareNode)])
+
+    measure_nodes_to_remove = [m for p in prepare_nodes_removed for m, p_ in cut_edges if p is p_]
+    communication_graph.remove_nodes_from(subgraphs_indices_to_remove)
+
+    for m in measure_nodes_to_remove:
+        for s in subgraphs_connected_to_measurements:
+            if s.has_node(m):
+                s.remove_node(m)
```

Comment:
The above code does:
- Find which subgraphs have terminal measurements. These subgraphs are enumerated by `terminal_indices`
- Loop through all subgraphs and check if there is a path in the communication graph to any of the terminal subgraphs
- If there is no path, the subgraph is disconnected. We need to remove it from the list of subgraphs and the node from the communication graph.
- We then remove any `MeasureNode` or `PrepareNode` operations from valid remaining subgraphs if those nodes connected to an invalid (disconnected) subgraph

Reply:
> Could the check be performed earlier in the pipeline e.g in replace_wire_cut_node so instead of adding and removing nodes, we identify the WireCut is redundant and we don't add the nodes at all?

I'm not sure how we could do it before the fragmentation stage, since we need to inspect the connectivity between subgraphs. One potential approach could be to "zoom in" and do the above process but for nodes in the circuit graph, i.e., finding which operations are not connected to terminal measurements. We already do this for gradient calculations (see [here](https://github.com/PennyLaneAI/pennylane/blob/master/pennylane/tape/jacobian_tape.py#L141)) but maybe it could be extended more generally.